File: scripts/axomeme/verify_preprocessing.py
#!/usr/bin/env python3
"""
Emit REFERENCE patristic distances for a set of newick trees, for the JS port to be checked against.

WHY THIS RUNS THE ML TEAM'S CODE RATHER THAN A TRANSCRIPTION OF IT.

A parity harness that compares our JS against our Python re-implementation proves only that we made
the same mistake twice. So this does not reimplement anything: it extracts the SOURCE TEXT of
`calculate_patristic_distances` out of the handoff's predict_regression_nexus.py and execs it. What
runs is byte-for-byte their function.

It is extracted rather than imported because importing that module pulls torch, pandas, scipy and
sklearn at module scope, none of which this comparison needs — and requiring a GPU-class dependency
tree to check a tree walk is how a parity gate stops being run.

The companion is verify_preprocessing.mjs, which consumes this output. Neither script writes anything
into the repository, and neither prints taxon names or sequence data: real DataMonkey submissions are
unpublished research, so only file ids, shapes and aggregate deltas are reported.

Usage:
  python3 scripts/axomeme/verify_preprocessing.py \
      --handoff /path/to/predict_regression_nexus.py \
      --trees   '/path/to/corpus/**/*.tre' \
      --out     /tmp/axomeme_reference.json
"""

import argparse
import glob
import json
import math
import re
import sys

REFERENCE_FNS = ["calculate_patristic_distances", "compute_mds_coordinates"]


def extract_reference(handoff_path, fn_name):
    """Pull one reference function's source out of the handoff, verbatim."""
    src = open(handoff_path, "r").read()
    start = src.find(f"def {fn_name}(")
    if start < 0:
        sys.exit(f"[!] {fn_name} not found in {handoff_path}")
    # Runs to the next top-level def/class.
    rest = src[start:]
    m = re.search(r"\n(?=(?:def |class )\w)", rest)
    return rest[: m.start()] if m else rest


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--handoff", required=True, help="path to predict_regression_nexus.py")
    ap.add_argument("--trees", required=True, help="glob for newick files")
    ap.add_argument("--out", required=True)
    ap.add_argument("--limit", type=int, default=0)
    ap.add_argument(
        "--max-species",
        type=int,
        default=512,
        help="pad the distance matrix to this size before MDS, as the driver does (default 512)",
    )
    args = ap.parse_args()

    from Bio import Phylo  # noqa: E402
    import numpy as np  # noqa: E402

    ns = {"math": math, "np": np}
    for fn in REFERENCE_FNS:
        body = extract_reference(args.handoff, fn)
        exec(compile(body, "<reference>", "exec"), ns)
        print(f"[*] Extracted {fn} ({body.count(chr(10))} lines) from {args.handoff}")
    reference = ns["calculate_patristic_distances"]
    reference_mds = ns["compute_mds_coordinates"]

    paths = sorted(glob.glob(args.trees, recursive=True))
    if args.limit:
        paths = paths[: args.limit]
    print(f"[*] {len(paths)} tree files")

    out, crashed, unreadable = {}, [], []
    for p in paths:
        try:
            tree = Phylo.read(p, "newick")
        except Exception as ex:
            unreadable.append((p, type(ex).__name__))
            continue
        try:
            names, dist, _ = reference(tree)
        except Exception as ex:
            # EXPECTED for a real slice of the corpus, and worth counting rather than skipping: the
            # reference computes log((node_count + 1.0) / (dist + 0.1)) unconditionally, so any
            # patristic distance <= -0.1 is a math domain error. DM3's own NJ emits the negative
            # branch lengths that produce those.
            crashed.append((p, type(ex).__name__, str(ex)[:80]))
            continue
        names = [n for n in names if n]
        # The driver pads the distance matrix to max_species BEFORE running MDS, so the padded zeros
        # take part in the double-centring and the coordinates depend on max_species. Reproduce that.
        # Trees larger than the cap would go through Max-PD selection in the driver; here they are
        # simply truncated, identically on both sides, because Max-PD is covered by unit tests and
        # mixing the two would make a failure ambiguous.
        names = names[: args.max_species]
        n = args.max_species
        padded = np.zeros((n, n), dtype=np.float32)
        for i, a in enumerate(names):
            row = dist.get(a, {})
            for j, b in enumerate(names):
                padded[i, j] = row.get(b, 0.0)
        coords = reference_mds(padded, n_components=4)
        out[p] = {
            "names": names,
            "dist": [[dist[a].get(b, 0.0) for b in names] for a in names],
            "max_species": n,
            "mds": [[float(v) for v in row] for row in coords],
        }

    json.dump(out, open(args.out, "w"))
    print(f"[*] wrote {len(out)} reference matrices to {args.out}")
    print(f"[*] unreadable by Bio.Phylo: {len(unreadable)}")
    print(f"[*] REFERENCE CRASHED on {len(crashed)} trees ({100.0 * len(crashed) / max(1, len(paths)):.1f}%)")
    for p, kind, msg in crashed[:5]:
        print(f"      {p.split('/')[-1]}: {kind}: {msg}")


if __name__ == "__main__":
    main()

File: scripts/axomeme/verify_preprocessing.mjs
/**
 * Compare the JS patristic-distance port against the ML team's own Python, over real trees.
 *
 * This is the gate that makes the port trustworthy. The unit tests in
 * src/test/axomeme-patristic.test.js check hand-computed values on trees small enough to verify by
 * eye; this checks the SAME CODE against the reference implementation on real DataMonkey
 * submissions, where the trees are large, unbalanced, occasionally malformed, and carry the negative
 * branch lengths DM3's own NJ produces.
 *
 * Run verify_preprocessing.py first to produce the reference JSON.
 *
 *   node scripts/axomeme/verify_preprocessing.mjs \
 *       --reference /tmp/axomeme_reference.json \
 *       [--tolerance 1e-9]
 *
 * Exits non-zero if any tree exceeds the tolerance, so it can be wired into CI the way
 * scripts/prescreen/verify_parity.py is. Prints file ids, shapes and deltas only — never taxon
 * names or sequence data, because the corpus is unpublished research data.
 */

import { readFileSync } from 'node:fs';
import {
	parseNewick,
	leafIndex,
	normalizeTaxonName
} from '../../src/lib/services/axomeme/newick.js';
import { patristicMatrix } from '../../src/lib/services/axomeme/patristic.js';
import { computeMdsCoordinates } from '../../src/lib/services/axomeme/mds.js';

const arg = (flag, fallback) => {
	const i = process.argv.indexOf(flag);
	return i > 0 && process.argv[i + 1] ? process.argv[i + 1] : fallback;
};

const referencePath = arg('--reference');
const tolerance = Number(arg('--tolerance', '1e-9'));
if (!referencePath) {
	console.error('usage: verify_preprocessing.mjs --reference <json> [--tolerance 1e-9]');
	process.exit(2);
}

const reference = JSON.parse(readFileSync(referencePath, 'utf8'));
const entries = Object.entries(reference);
console.log(`[*] ${entries.length} reference matrices, tolerance ${tolerance}`);

let checked = 0;
let cells = 0;
let worst = 0;
let worstFile = null;
const failures = [];
const skipped = [];

/**
 * MDS gets its own, much looser tolerance, and the reason is not slack.
 *
 * The coordinates are cast to float32 by the reference, so ~1e-7 relative is the floor no matter how
 * carefully either side computes. On top of that, this is the one stage where two CORRECT
 * implementations can legitimately disagree: numpy uses divide-and-conquer (dsyevd), this port uses
 * implicit-shift QL, and inside a degenerate eigenspace any orthonormal basis is a valid answer that
 * the reference's sign convention cannot disambiguate. So the number below is a measurement
 * threshold, not a correctness proof — what matters is the DISTRIBUTION printed at the end.
 */
const mdsTolerance = Number(arg('--mds-tolerance', '1e-5'));
let mdsChecked = 0;
let mdsCells = 0;
let mdsWorstOverall = 0;
let mdsWorstFile = null;
const mdsFailures = [];

for (const [path, ref] of entries) {
	const id = path.split('/').pop();
	let tree;
	try {
		tree = parseNewick(readFileSync(path, 'utf8'));
	} catch (e) {
		skipped.push(`${id}: parse threw ${e.message}`);
		continue;
	}

	const { index } = leafIndex(tree);
	// Order the rows exactly as the reference did, so cell (i,j) means the same pair on both sides.
	const nodes = ref.names.map((n) => index.get(normalizeTaxonName(n)));
	if (nodes.some((v) => v === undefined)) {
		// A name the JS parser did not produce is a PARSE divergence, not a distance divergence, and
		// it is a real failure — it means the two parsers disagree about what the taxa are.
		const missing = nodes.filter((v) => v === undefined).length;
		failures.push(`${id}: ${missing}/${ref.names.length} taxa missing from the JS parse`);
		continue;
	}

	const n = nodes.length;
	const mine = patristicMatrix(tree, nodes);
	let fileWorst = 0;
	for (let i = 0; i < n; i++) {
		for (let j = 0; j < n; j++) {
			const d = Math.abs(mine[i * n + j] - ref.dist[i][j]);
			if (d > fileWorst) fileWorst = d;
		}
	}
	cells += n * n;
	checked++;
	if (fileWorst > worst) {
		worst = fileWorst;
		worstFile = `${id} (${n} taxa)`;
	}
	if (fileWorst > tolerance) {
		failures.push(`${id}: ${n} taxa, max|Δ| ${fileWorst.toExponential(3)}`);
	}

	// --- MDS, the piece whose agreement is not guaranteed by construction ---
	if (!ref.mds) continue;
	const cap = ref.max_species;
	// The reference pads to max_species BEFORE the eigendecomposition, so the padded zeros take part
	// in the double-centring. Build the same padded matrix rather than running MDS on the real taxa.
	const padded = new Float64Array(cap * cap);
	for (let i = 0; i < n; i++) {
		for (let j = 0; j < n; j++) padded[i * cap + j] = mine[i * n + j];
	}
	const coords = computeMdsCoordinates(padded, cap, 4);
	let mdsWorst = 0;
	let mdsWorstComponent = -1;
	let mdsScale = 0;
	for (let i = 0; i < cap; i++) {
		for (let c = 0; c < 4; c++) {
			const d = Math.abs(coords[i * 4 + c] - ref.mds[i][c]);
			if (d > mdsWorst) {
				mdsWorst = d;
				mdsWorstComponent = c;
			}
			const s = Math.abs(ref.mds[i][c]);
			if (s > mdsScale) mdsScale = s;
		}
	}
	// GATE ON RELATIVE ERROR, against the largest coordinate in the whole matrix.
	//
	// An absolute threshold is the wrong instrument here and reads as a failure when nothing is wrong.
	// The four components have wildly different scales — a measured 135-taxon tree runs 9.3e2, 2.2e2,
	// 6.4e0, 2.1e-1 — because they carry eigenvalues seven orders of magnitude apart. A fixed absolute
	// tolerance is simultaneously far too loose for component 0 and far too tight for component 3.
	// What the model actually consumes is the vector as a whole, through one Linear layer, so error
	// relative to the vector's own scale is the quantity that means something.
	const mdsRel = mdsScale > 0 ? mdsWorst / mdsScale : 0;
	mdsChecked++;
	mdsCells += cap * 4;
	if (mdsRel > mdsWorstOverall) {
		mdsWorstOverall = mdsRel;
		mdsWorstFile = `${id} (${n} taxa, component ${mdsWorstComponent}, abs ${mdsWorst.toExponential(2)})`;
	}
	if (mdsRel > mdsTolerance) {
		mdsFailures.push(
			`${id}: ${n} taxa -> ${cap}, rel ${mdsRel.toExponential(3)} ` +
				`(abs ${mdsWorst.toExponential(3)}, scale ${mdsScale.toExponential(3)}) on component ${mdsWorstComponent}`
		);
	}
}

console.log(`\n--- patristic distances ---`);
console.log(`[*] compared ${checked} trees, ${cells.toLocaleString()} distance cells`);
console.log(`[*] worst |Δ| ${worst.toExponential(3)}${worstFile ? `  (${worstFile})` : ''}`);
if (skipped.length) {
	console.log(`[!] ${skipped.length} skipped:`);
	for (const s of skipped.slice(0, 5)) console.log(`      ${s}`);
}

if (mdsChecked) {
	console.log(`\n--- MDS coordinates ---`);
	console.log(`[*] compared ${mdsChecked} trees, ${mdsCells.toLocaleString()} coordinates`);
	console.log(
		`[*] worst RELATIVE Δ ${mdsWorstOverall.toExponential(3)}${mdsWorstFile ? `  (${mdsWorstFile})` : ''}`
	);
	console.log(
		`[*] float32 relative resolution is ~1.2e-7, so anything near that is the format, not the port`
	);
	console.log(
		`[*] ${mdsChecked - mdsFailures.length}/${mdsChecked} trees within ${mdsTolerance}` +
			` (${(((mdsChecked - mdsFailures.length) / mdsChecked) * 100).toFixed(1)}%)`
	);
	if (mdsFailures.length) {
		console.log(`[!] ${mdsFailures.length} over tolerance:`);
		for (const f of mdsFailures.slice(0, 20)) console.log(`      ${f}`);
	}
}

if (failures.length) {
	console.log(`\n[!] ${failures.length} DISTANCE FAILURES:`);
	for (const f of failures.slice(0, 20)) console.log(`      ${f}`);
	process.exit(1);
}
if (mdsFailures.length) {
	console.log('\n[!] MDS PARITY NOT CLEAN');
	process.exit(1);
}
console.log('\n[*] PARITY OK');
